creepy/query/__init__.py

```diff
@@ -0,0 +1,195 @@
+
+    def __delattr__(self, name):
+        self._remote._post(DelattrQuery(self._id, name))
+
+    def __call__(self, *args, **kwargs):
+        return self._make_child(CallQuery(self._id, args, kwargs))
+
+    def _catch_magic_call(name):
+        def handler(self, *args, **kwargs):
+            return self.__getattr__(name)(*args, **kwargs)
+        return handler
+
+    def _catch_magic_call_nd_download(name):
+        def handler(self, *args, **kwargs):
+            return self._remote.download(self.__getattr__(name)(*args, **kwargs))
+        return handler
+
+    __enter__ = _catch_magic_call('__enter__')
+    __exit__ = _catch_magic_call('__exit__')
+    __bool__ = _catch_magic_call_nd_download('__bool__')
+    __str__ = _catch_magic_call_nd_download('__str__')
+    __repr__ = _catch_magic_call_nd_download('__repr__')
+
+
+def _make_request(url, data=None):
+    response = requests.post(url, data)
+    if response.status_code == 200:
+        return response.content
+    logger.error(response.content)
+    return None
+
+
+class Remote:
+    def __init__(self, url, session_id, cipher):
+        self._url = url
+        self._session_id = session_id
+        self._cipher = cipher
+        self._nonce = 0
+
+    @property
+    def globals(self):
+        return ProxyObject(self, 0)
+
+    def _post(self, query):
+        data = self._nonce.to_bytes(NONCE_SIZE, 'big') + pickle.dumps(query, PICKLE_PROTOCOL)
+        self._nonce += 1
+        response = _make_request(self._url, self._session_id + self._cipher.encrypt(data))
+        if response is None:
+            raise ValueError()
+        res = pickle.loads(self._cipher.decrypt(response))
+        if isinstance(res, Exception):
+            raise res
+        return res
+
+    def download(self, obj: ProxyObject):
+        assert self == obj._remote, 'The object is on a different node'
+        return self._post(DownloadQuery(obj._id))
+
+    def _send_file(self, src_path: str, dst_path: str, exist_ok=False):
+        CHUNK_SIZE = 2**17
+        if not exist_ok and self.globals.os.path.exists(dst_path):
+            raise OSError(f"File exists: '{dst_path}'")
+        with self.globals.open(dst_path, 'wb') as dst_f:
+            with open(src_path, 'rb') as src_f:
+                while True:
+                    chunk = src_f.read(CHUNK_SIZE)
+                    if not chunk:
+                        break
+                    dst_f.write(chunk)
+
+    def _send_directory(self, src_dir: str, dst_dir: str, exist_ok=False):
+        remote_os = self.globals.os
+        remote_os.makedirs(dst_dir, exist_ok=exist_ok)
+        for src_root, dirs, files in os.walk(src_dir):
+            dst_root = os.path.join(dst_dir, os.path.relpath(src_root, src_dir))
+            for name in files:
+                self._send_file(os.path.join(src_root, name), os.path.join(dst_root, name))
+            for name in dirs:
+                remote_os.makedirs(os.path.join(dst_root, name), exist_ok=exist_ok)
```

Comment:
Shouldn't you call `_send_directory` here?

Reply:
os.walk traverses directories recursively.

Reply:
So then change order of loops. First create subdirectories and then send files.

Reply:
Order doesn't matter, it enumerate file names and dir names in the same directory that already exists on destination.

Reply:
Don't get it. Why did it exist on the destination?

Reply:
They are created in lines 168 and 174.